trace/client/src/main/scala/org/http4s/otel4s/middleware/trace/client/ClientMiddlewareBuilder.scala

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 2023 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package otel4s.middleware
+package trace
+package client
+
+import cats.effect.Concurrent
+import cats.effect.MonadCancelThrow
+import cats.effect.Outcome
+import cats.effect.Resource
+import cats.syntax.applicative._
+import cats.syntax.flatMap._
+import fs2.Stream
+import org.http4s.client.Client
+import org.typelevel.otel4s.trace.SpanKind
+import org.typelevel.otel4s.trace.StatusCode
+import org.typelevel.otel4s.trace.Tracer
+
+/** Middleware builder for wrapping an http4s `Client` to add tracing.
+  *
+  * @see [[https://opentelemetry.io/docs/specs/semconv/http/http-spans/#http-client]]
+  */
+class ClientMiddlewareBuilder[F[_]: Tracer: Concurrent] private (
```

Comment:
Another point worth discussing, since it wasn't outlined anywhere, is that the **instrumentation** library should require `TracerProvider` instead of `Tracer`.

Some context: https://opentelemetry.io/docs/concepts/instrumentation/libraries/#getting-a-tracer.

Ideally, we must use a **scoped** variant of `Tracer` and `Meter` rather than a global one from the application entry point. 

That way, we can help organize the telemetry signals into logical groups.

```scala
TracerProvider[F].tracer("org.http4s.client").withVersion(BuildInfo.version).get
  .flatMap { implicit tracer =>
    // rest of the code
  }
```
_____

I don't think we should act immediately, but it would be nice to introduce instrumentation guidelines in the otel4s repo so it would be easier for everyone.

Reply:
I think I'll take care of that in a separate PR. relatedly, it might also be worth adding an attribute to the spans with the middleware version?

Reply:
bikeshedding: what should the name of the tracer we try to get be? how precise should it be? we can go with `org.http4s.otel4s.middleware.{client,server}`, or we could go with something as short as your example. do we worry about what name potential future instrumentations directly in http4s might use?

Reply:
> I think I'll take care of that in a separate PR. 

Yeah, let's don't rush it. 

>  we can go with org.http4s.otel4s.middleware.{client,server}, or we could go with something as short as your example.

I don't have an answer yet 😅 Here is an example: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/57c7cf2ad5f7272c1eb83b9816e652bf832c91d4/instrumentation/lettuce/lettuce-5.1/library/src/main/java/io/opentelemetry/instrumentation/lettuce/v5_1/LettuceTelemetry.java#L18

I think `org.http4s.otel4s.middleware.{client,server}` fits better. Especially since we are going to use otel4s in http4s eventually. 

> do we worry about what name potential future instrumentations directly in http4s might use? 

AFAIK we shouldn't use signal (e.g. trace, metric, etc) suffix, so
`TracerProvider[F].tracer("org.http4s.otel4s.middleware.client")` and `MeterProvider[F].meter("org.http4s.otel4s.middleware.client")` should be ok.

Reply:
was waiting on reviews, so I decided to put it in this PR anyway